giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/tokens/impl/IntrospectTokenService.java

```diff
@@ -0,0 +1,157 @@
+import org.springframework.stereotype.Service;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestTemplate;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.nimbusds.jose.util.Base64;
+
+import edu.asu.diging.gilesecosystem.util.properties.IPropertiesManager;
+import edu.asu.diging.gilesecosystem.web.apps.IRegisteredApp;
+import edu.asu.diging.gilesecosystem.web.exceptions.AppMisconfigurationException;
+import edu.asu.diging.gilesecosystem.web.service.apps.IRegisteredAppManager;
+import edu.asu.diging.gilesecosystem.web.service.properties.Properties;
+import edu.asu.diging.gilesecosystem.web.tokens.IApiTokenContents;
+import edu.asu.diging.gilesecosystem.web.tokens.IIntrospectTokenService;
+
+/**
+ * Class to introspect access token for MITREid connect server
+ * @author snilapwa
+ *
+ */
+@Service
+public class IntrospectTokenService implements IIntrospectTokenService{
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Autowired
+    private IPropertiesManager propertyManager;
+
+    @Autowired
+    private IRegisteredAppManager appsManager;
+
+    private HttpComponentsClientHttpRequestFactory factory;
+
+    public IntrospectTokenService() {
+        this(HttpClientBuilder.create().useSystemProperties().build());
+    }
+
+    public IntrospectTokenService(HttpClient httpClient) {
+        this.factory = new HttpComponentsClientHttpRequestFactory(httpClient);
+    }
+
+    /**
+     * Calls MITREid connect server introspect api using client with
+     * protected resource access.
+     *  @param accessToken accessToken to pass to introspect Api
+     *  @param appId appId to get provider client
+     *  @return tokenContents with username and expiration status
+     */
+    public IApiTokenContents introspectAccessToken(String accessToken, String appId) throws AppMisconfigurationException {
+
+        IRegisteredApp app = appsManager.getApp(appId);
+
+        if (app.getProviderClientId() == null || app.getProviderClientId().isEmpty()) {
+            throw new AppMisconfigurationException("No provider client id has been registered for your app.");
+        }
+
+        // use mitreid connect client with protected resource access
+        final String clientId = propertyManager.getProperty(Properties.MITREID_INTROSPECT_CLIENT_ID);
+        final String clientSecret = propertyManager.getProperty(Properties.MITREID_INTROSPECT_SECRET);
+
+        String introspectionUrl = propertyManager.getProperty(Properties.MITREID_INTROSPECT_URL);
+
+        if (introspectionUrl == null || introspectionUrl.isEmpty()) {
+            throw new AppMisconfigurationException("Unable to load server URL");
```

Comment:
this should probably throw a server misconfiguration exception, given that the server isnt' configured with the URL for the introspect endpoint.